# SQL Practice Case
## IYKRA Data Fellowship Batch 5 2021
### Abednego Kristanto

Saya mencoba untuk mengerjakan practice case SQL pada IYKRA Data Fellowship Batch 5 2021 di Jupyter Notebook pada PC saya. Setelah mencari beberapa cara untuk menghubungkan database PostgreSQL dengan Jupyter Notebook, akhirnya saya menemukan salah satu cara yang berhasil saya terapkan yaitu menggunakan package SQLAlchemy.

In [1]:
!pip install sqlalchemy

In [2]:
import sqlalchemy
import pandas as pd
import postgres_access as access
from sqlalchemy import create_engine

Fungsi dari create_engine adalah membuat sebuah obyek engine yang berbasis URL untuk menghubungkan database PostgreSQL dengan Jupyter notebook. Database URL yang dibutuhkan oleh engine ini biasanya ditulis sebagai string dengan format sebagai berikut:

"dialect+driver://username:password@host_address:port/database_name"

Data-data yang bersifat confidential seperti username dan password disini dapat dimasukkan ke dalam file Python tersendiri yang hanya bisa diakses oleh user yang bersangkutan. Disini saya membuat sebuah file Python sederhana yang disimpan di folder yang sama dengan notebook ini di PC saya dengan nama postgress_access.py. Isi dari postgress_access.py adalah sebagai berikut:  

POSTGRES_USERNAME = "postgersql username"  
POSTGRES_PASSWORD = "postgersql password"  
POSTGRES_ADDRESS = "postgersql address"  
POSTGRES_PORT = "postgersql port"  

In [3]:
postgres_str = ('postgresql://{username}:{password}@{ipaddress}:{port}/{dbname}'
                .format(username=access.POSTGRES_USERNAME,
                        password=access.POSTGRES_PASSWORD,
                        ipaddress=access.POSTGRES_ADDRESS,
                        port = access.POSTGRES_PORT,
                        dbname = 'dvdrental'))

engine = create_engine(postgres_str)

Setelah engine berhasil dibuat, engine ini akan menjalankan SQL query yang ditulis dalam bentuk string. Metode pandas read_sql_query bisa digunakan untuk langsung menyimpan hasil SQL query menjadi pandas DataFrame untuk selanjutnya dapat diolah lebih lanjut.

### Questions:
Berikut ini adalah pertanyaan-pertanyaan pada Practice Case SQL IYKRA Data Fellowship Batch 5 2021 yang saya coba kerjakan di Jupyter Notebook.
#### 1. A customer wants to know the films about ‘ancient’. How many recommendations could you give for him?

In [4]:
str_query1a = "\
SELECT COUNT(title) AS title_num \
FROM film \
WHERE fulltext @@ to_tsquery('ancient’'); \
"

In [5]:
pd.read_sql_query(str_query1a, engine)

title_num
0        124

In [6]:
str_query1b = "\
SELECT title, fulltext FROM film \
WHERE fulltext @@ to_tsquery('ancient’'); \
"

In [7]:
result1 = pd.read_sql_query(str_query1b, engine)
result1.head(5)

title                                           fulltext
0     Airport Pollock  'airport':1 'ancient':18 'confront':14 'epic':...
1      Ace Goldfinger  'ace':1 'administr':9 'ancient':19 'astound':4...
2        Agent Truman  'agent':1 'ancient':19 'boy':11 'china':20 'es...
3    Aladdin Calendar  'action':5 'action-pack':4 'aladdin':1 'ancien...
4  Antitrust Tomatoes  'administr':17 'ancient':19 'antitrust':1 'dat...

In [8]:
result1['fulltext'][4]

"'administr':17 'ancient':19 'antitrust':1 'databas':16 'fate':4 'feminist':11 'india':20 'must':13 'succumb':14 'tomato':2 'woman':8 'yarn':5"

There are 124 record in the database about films that have 'Ancient' in its description.

#### 2. I wonder, how many films have a rating of "R" and a replacement cost between \\$5 and \\$10?

In [9]:
str_query2a = "\
SELECT COUNT(film_id) AS num_film \
FROM film \
WHERE rating = 'R' \
AND replacement_cost BETWEEN 5 AND 10; \
"

In [10]:
pd.read_sql_query(str_query2a, engine)

num_film
0         7

In [11]:
str_query2b = "\
SELECT title, rating, replacement_cost FROM film \
WHERE rating = 'R' \
AND replacement_cost BETWEEN 5 AND 10; \
"

In [12]:
result2 = pd.read_sql_query(str_query2b, engine)
result2

title rating  replacement_cost
0    Anaconda Confessions      R              9.99
1  Deliverance Mulholland      R              9.99
2           Kissing Dolls      R              9.99
3          Pluto Oleander      R              9.99
4            Purple Movie      R              9.99
5           Roxanne Rebel      R              9.99
6         Sun Confessions      R              9.99

There are 7 films with rating 'R' and replacement cost between \\$5 and \\$10

#### 3. We have two staff members with staff IDs 1 and 2. We want to give a bonus to the staff member that handled the most payments. How many payments did each staff member handle? And how much was the total amount processed by each staff member

In [13]:
str_query3 = "\
SELECT s.staff_id, COUNT(payment_id) AS num_payment, \
    SUM(amount) AS total_amount \
FROM staff AS s \
INNER JOIN payment as p \
    ON s.staff_id = p.staff_id \
GROUP BY s.staff_id \
ORDER BY COUNT(payment_id) DESC; \
"

In [14]:
pd.read_sql_query(str_query3, engine, index_col='staff_id')

num_payment  total_amount
staff_id                           
2                7304      31059.92
1                7292      30252.12

Staff ID 2 handle 7304 number of payments, and staff ID 2 handle 7292 number of payments. Total amount of payments handled is \\$31059.92, and \\$30353.12 is handled by staff ID 2, and staff ID 1, respectively. 

#### 4. Corporate headquarters is auditing the store, they want to know the average replacement cost of movies by rating

In [15]:
str_query4 = "\
SELECT rating, \
    AVG(replacement_cost) AS avg_replace_cost \
FROM film \
GROUP BY rating; \
"

In [16]:
pd.read_sql_query(str_query4, engine)

rating  avg_replace_cost
0  NC-17         20.137619
1      G         20.124831
2  PG-13         20.402556
3      R         20.231026
4     PG         18.959072

#### 5. We want to send coupons to the 5 customers who have spent the most amount of money. Get the customer name, emal and their spent amount!

In [17]:
str_query5 = "\
SELECT first_name, last_name, email, \
    SUM(amount) AS total_amount \
FROM customer AS c \
INNER JOIN payment AS p \
    ON c.customer_id = p.customer_id \
GROUP BY c.customer_id \
ORDER BY SUM(amount) DESC \
LIMIT 5; \
"

In [18]:
pd.read_sql_query(str_query5, engine)

first_name last_name                              email  total_amount
0    Eleanor      Hunt    eleanor.hunt@sakilacustomer.org        211.55
1       Karl      Seal       karl.seal@sakilacustomer.org        208.58
2     Marion    Snyder   marion.snyder@sakilacustomer.org        194.61
3     Rhonda   Kennedy  rhonda.kennedy@sakilacustomer.org        191.62
4      Clara      Shaw      clara.shaw@sakilacustomer.org        189.60

#### 6. I wonder, how many distinct users have rented each genre? And what is the average rental rate per movie genre? (from the highest to the lowest rental rate)

In [19]:
str_query6 = "\
SELECT name AS genre, COUNT(DISTINCT(customer_id)) AS num_customers, \
    AVG(rental_rate) AS avg_rent_rate \
FROM category AS c, film_category AS fc, \
    film AS f, inventory AS i, rental AS r \
WHERE c.category_id = fc.category_id AND \
    fc.film_id = f.film_id AND \
    f.film_id = i.film_id AND \
    i.inventory_id = r.inventory_id \
GROUP BY name \
ORDER BY AVG(rental_rate) DESC; \
"

In [20]:
pd.read_sql_query(str_query6, engine)

genre  num_customers  avg_rent_rate
0        Travel            442       3.317360
1        Comedy            495       3.283305
2         Drama            501       3.182453
3         Games            474       3.130351
4        Horror            451       3.101111
5           New            468       3.090000
6        Sports            519       3.068032
7         Music            447       3.062289
8        Sci-Fi            507       2.988183
9       Foreign            493       2.953214
10    Animation            500       2.760154
11       Family            501       2.699854
12     Children            482       2.689471
13       Action            510       2.668058
14     Classics            468       2.638562
15  Documentary            483       2.616667

#### 7. We want to know what customers are eligible for our platinum credit card. The requirements are that the customer has at least a tote of 40 transaction payments. Get the customer name, email who are eligible for the credit card!

In [21]:
str_query7 = "\
SELECT first_name, last_name, email, \
    COUNT(payment_id) AS total_payments \
FROM payment AS p \
INNER JOIN customer as c \
    ON p.customer_id = c.customer_id \
GROUP BY p.customer_id, first_name, last_name, email \
HAVING COUNT(payment_id) >= 40 \
ORDER BY COUNT(payment_id) DESC; \
"

In [22]:
pd.read_sql_query(str_query7, engine)

first_name last_name                            email  total_payments
0    Eleanor      Hunt  eleanor.hunt@sakilacustomer.org              45
1       Karl      Seal     karl.seal@sakilacustomer.org              42
2      Clara      Shaw    clara.shaw@sakilacustomer.org              40